integrations/wiz/client.py

```diff
@@ -0,0 +1,246 @@
+    ):
+        self.api_url = api_url
+        self.client_id = client_id
+        self.client_secret = client_secret
+        self.token_url = token_url
+        self.last_token_object: TokenResponse | None = None
+
+        self.http_client = http_async_client
+
+    @property
+    async def api_auth_headers(self) -> dict[str, Any]:
+        token = self._get_token()
+
+        return {
+            "Authorization": token.full_token,
+            "Content-Type": "application/json",
+        }
+
+    @property
+    def auth_params(self) -> dict[str, Any]:
+        audience_mapping = {
+            **{url: "beyond-api" for url in AUTH0_URLS},
+            **{url: "wiz-api" for url in COGNITO_URLS},
+        }
+
+        if self.token_url not in audience_mapping:
+            raise Exception("Invalid Token URL")
+
+        audience = audience_mapping[self.token_url]
+
+        return {
+            "grant_type": "client_credentials",
+            "audience": audience,
+            "client_id": self.client_id,
+            "client_secret": self.client_secret,
+        }
+
+    def _get_token(self) -> TokenResponse:
+        try:
+            response = httpx.post(
+                self.token_url,
+                data=self.auth_params,
+                headers={"Content-Type": "application/x-www-form-urlencoded"},
+            )
+            response.raise_for_status()
+            return TokenResponse(**response.json())
+        except Exception as e:
+            logger.exception(e)
+            raise
+
+    @property
+    async def token(self) -> str:
+        if not self.last_token_object or not self.last_token_object.is_valid:
+            msg = "Wiz Token expired or invalid, fetching new token"
+            logger.info(msg)
+            self.last_token_object = await self._get_token(
+                self.client_id, self.client_secret
+            )
+
+        return self.last_token_object.token
+
+    async def get_issues(self) -> AsyncGenerator[list[dict[str, Any]], None]:
+        logger.info("Fetching issues from Wiz API")
+
+        if cache := event.attributes.get(CacheKeys.ISSUES):
+            logger.info("picking Wiz issues from cache")
+            yield cache
+
+        json_data: dict[str, Any] = {
+            "query": ISSUES_GQL,
+            "variables": {
+                "first": PAGE_SIZE,
+                "orderBy": {"direction": "DESC", "field": "CREATED_AT"},
+            },
+        }
+
+        try:
+            response = await self.http_client.post(
+                url=self.api_url,
+                json=json_data,
+                headers=await self.api_auth_headers,
+            )
+            response.raise_for_status()
+            response_json = response.json()
+            data = response_json.get("data")
+
+            if not data:
+                logger.error(response_json.get("error"))
+                raise Exception("No data found for Wiz account")
```

Comment:
is there a case the status code will be ok and the response will have error?

Reply:
There is a possibility of an unexpected internal server error happening and they return 200 with no data.


Error | Potential issues
-- | --
200 UNAUTHORIZED | The Wiz service account token scopes do not permit this action. Check the error message for the required scopes.
200 INTERNAL | An unexpected internal server error. Contact the Wiz Integrations team.